test/cmp_protect_test.c

```diff
@@ -0,0 +1,535 @@
+    EVP_PKEY *pubkey;
+    unsigned char *mem;
+    int memlen;
+    X509 *cert;
+    STACK_OF(X509) *certs;
+    STACK_OF(X509) *chain;
+    int callback_arg;
+    int expected;
+} CMP_PROTECT_TEST_FIXTURE;
+
+static void tear_down(CMP_PROTECT_TEST_FIXTURE *fixture)
+{
+    /* ERR_print_errors_fp(stderr);
+       Free any memory owned by the fixture, etc. */
+    OSSL_CMP_CTX_free(fixture->cmp_ctx);
+    OSSL_CMP_MSG_free(fixture->msg);
+    ASN1_OCTET_STRING_free(fixture->secret);
+    EVP_PKEY_free(fixture->privkey);
+    EVP_PKEY_free(fixture->pubkey);
+    OSSL_CMP_PKISI_free(fixture->si);
+
+    OPENSSL_free(fixture->mem);
+    sk_X509_free(fixture->certs);
+    sk_X509_free(fixture->chain);
+
+    OPENSSL_free(fixture);
+}
+
+static CMP_PROTECT_TEST_FIXTURE *set_up(const char *const test_case_name)
+{
+    CMP_PROTECT_TEST_FIXTURE *fixture;
+
+    if (!TEST_ptr(fixture = OPENSSL_zalloc(sizeof(*fixture))))
+        return NULL;
+    fixture->test_case_name = test_case_name;
+    if (!TEST_ptr(fixture->cmp_ctx = OSSL_CMP_CTX_new())) {
+        tear_down(fixture);
+        return NULL;
+    }
+    return fixture;
+}
+
+static EVP_PKEY *loadedprivkey = NULL;
+static EVP_PKEY *loadedpubkey = NULL;
+static EVP_PKEY *loadedkey = NULL;
+static X509 *cert = NULL;
+static unsigned char rand_data[OSSL_CMP_TRANSACTIONID_LENGTH];
+static OSSL_CMP_MSG *ir_unprotected, *ir_protected;
+static X509 *endentity1 = NULL, *endentity2 = NULL,
+    *root = NULL, *intermediate = NULL;
+
+static int execute_calc_protection_fails_test(CMP_PROTECT_TEST_FIXTURE *fixture)
+{
+    ASN1_BIT_STRING *protection =
+        ossl_cmp_calc_protection(fixture->msg, fixture->secret, fixture->privkey);
+    int res = TEST_ptr_null(protection);
+
+    ASN1_BIT_STRING_free(protection);
+    return res;
+}
+
+/* TODO internal test*/
+static int execute_calc_protection_test(CMP_PROTECT_TEST_FIXTURE *fixture)
+{
+    ASN1_BIT_STRING *protection =
+        ossl_cmp_calc_protection(fixture->msg, fixture->secret, fixture->privkey);
+    int res = TEST_ptr(protection)
+                  && TEST_true(ASN1_STRING_cmp(protection,
+                                               fixture->msg->protection) == 0);
+
+    ASN1_BIT_STRING_free(protection);
+    return res;
+}
+
+/* This function works similar to parts of CMP_verify_signature in cmp_vfy.c,
+ * but without the need for a OSSL_CMP_CTX or a X509 certificate */
+static int verify_signature(OSSL_CMP_MSG *msg,
+                            ASN1_BIT_STRING *protection,
+                            EVP_PKEY *pkey, int digest_nid)
+{
+    CMP_PROTECTEDPART prot_part;
+    unsigned char *prot_part_der = NULL;
+    int l;
+    EVP_MD_CTX *ctx = NULL;
+    int res;
+
+    prot_part.header = OSSL_CMP_MSG_get0_header(msg);
+    prot_part.body = msg->body;
+    res =
+        TEST_int_ge(l = i2d_CMP_PROTECTEDPART(&prot_part, &prot_part_der), 0)
+            && TEST_ptr(ctx = EVP_MD_CTX_create())
+            && TEST_true(EVP_VerifyInit_ex
```

Comment:
Can we use `EVP_DigestVerify*` here instead. EVP_Verify* is considered legacy (we should probably deprecate it).

Reply:
I changed this.